bokeh/models/tools.py

```diff
@@ -865,3 +872,230 @@ class RedoTool(Action):
         :height: 18pt
 
     '''
+
+@abstract
+class EditTool(Tool):
+    ''' A base class for all interactive draw tool types.
+
+    '''
+
+    empty_value = Either(Bool, Int, Float, Date, Datetime, Color, help="""
+    Defines the value to insert on non-coordinate columns when a new
+    glyph is inserted into the ColumnDataSource columns, e.g. when a
+    circle glyph defines 'x', 'y' and 'color' columns, adding a new
+    point will add the x and y-coordinates to 'x' and 'y' columns and
+    the color column will be filled with the defined empty value.
+    """)
+
+    renderers = List(Instance(Renderer), help="""
+    An explicit list of renderers corresponding to scatter glyphs
+    that may be edited.
+    """)
+
+class BoxDrawTool(EditTool, Drag, Tap):
+    ''' *toolbar icon*: |box_draw_icon|
+
+    The BoxDrawTool allows drawing, dragging and deleting ``Rect``
+    glyphs on one or more renderers by editing the underlying
+    ``ColumnDataSource`` data. Like other drawing tools, the renderers
+    that are to be edited must be supplied explicitly as a list. When
+    drawing a new box the data will always be added to the
+    ``ColumnDataSource`` on the first supplied renderer.
+
+    The tool will automatically modify the columns on the data source
+    corresponding to the ``x``, ``y``, ``width`` and ``height`` values
+    of the glyph. Any additional columns in the data source will be
+    padded with the declared ``empty_value``, when adding a new box.
+
+    The supported actions include:
+
+    * Add box: Hold shift then click and drag anywhere on the plot.
+
+    * Move box: Click and drag an existing box, the box will be
+      dropped once you let go of the mouse button.
+
+    * Delete box: Tap a box to select it then press <<backspace>> key
+      while the mouse is within the plot area.
+
+    To **Move** or **Delete** multiple boxes at once:
+
+    * Move selection: Select box(es) with <<shift>>+tap (or another
+      selection tool) then drag anywhere on the plot. Selecting and
+      then dragging on a specific box will move both.
+
+    * Delete selection: Select box(es) with <<shift>>+tap (or another
+      selection tool) then press <<backspace>> while the mouse is
+      within the plot area.
+
+    .. |box_draw_icon| image:: /_images/icons/BoxDraw.png
+        :height: 18pt
+    '''
+
+    dimensions = Enum(Dimensions, default="both", help="""
+    Which dimensions the box drawing is to be free in. By default,
+    users may freely draw boxes with any dimensions. If only "width"
+    is supplied, the box will be constrained to span the entire
+    vertical space of the plot, only the horizontal dimension can be
+    controlled. If only "height" is supplied, the box will be
+    constrained to span the entire horizontal space of the plot, and
+    the vertical dimension can be controlled.
+    """)
```

Comment:
@philippjfr suddenly wonder if we should call this `BoxEditTool` to leave open the future possibility of additional modes to edit existing boxes (by dragging their edges to expend contract, etc)

Reply:
As it is this tool is *almost* useful for a somewhat often asked for capability: having a "zoom window" on one plot that controls the range on another.

Reply:
Sounds good to me, I did consider adding that while adding drag capability.

Reply:
OK well I think after that change we merge? And get a dev build out

Reply:
You mean adding the expand/contract mode? Only way I can currently think of is adding a vertex_renderer just like the PolyEditTool. Does that seem sensible?

Reply:
No, just changing the name now, to leave open the possibility of adding the editing features later without needing a new separate tool.